tests/fetch_service.rs

```diff
@@ -0,0 +1,26 @@
+#[cfg(feature = "wasm_test")]
+use wasm_bindgen_test::{wasm_bindgen_test as test, wasm_bindgen_test_configure};
+use yew::callback::Callback;
+use yew::format::Nothing;
+use yew::services::fetch::{self, FetchOptions, Response};
+use yew::services::FetchService;
+use yew::services::Task;
+
+#[cfg(feature = "wasm_test")]
+wasm_bindgen_test_configure!(run_in_browser);
+
+#[test]
+fn fetch_redirect_follow() {
+  let request = fetch::Request::get("https://httpbin.org/relative-redirect/1")
+    .body(Nothing)
+    .unwrap();
+  let options = FetchOptions {
+    redirect: Some(fetch::Redirect::Follow),
+    ..FetchOptions::default()
+  };
+  let callback = Callback::from(move |resp: Response<Result<String, failure::Error>>| {
+    println!("{:#?}", resp);
+  });
+  let task = FetchService::new().fetch_with_options(request, options, callback);
+  while task.is_active() {}
```

Comment:
How can I force the FetchTask to execute here? Looping doesnt do it.

Reply:
I would think that should execute without the `while`. I assume you aren't seeing a print in the console because a shim between rust's println and the browser console isn't set up, so by default, it should do nothing (although I may be wrong).

I don't recall how to do that with println, but https://github.com/yewstack/web_logger does this for the logging facilities in the `log` crate.

I may be entirely off-base though, and it isn't making the request at all.

The plain `fetch` is used here in this example: https://github.com/yewstack/yew/blob/master/examples/dashboard/src/lib.rs#L117

Reply:
Okay I see, I would think that without the loop, FetchTask would get dropped and immediately canceled.

In the example you linked, the FetchTask is returned and stored then, it does not get dropped.
I do not know how I can do that in the context of a test.
Also, I figured I could maybe use stdweb's support for PromiseFutures (https://github.com/koute/stdweb/pull/103) and convert the fetch "handle" to a future I can await but somehow there's nothing about them in stdweb's released documentation? https://docs.rs/stdweb/0.4.20/stdweb/?search=future https://docs.rs/stdweb/0.4.20/stdweb/?search=promise

Reply:
Ah, I forgot about that - and the example I provided looks suspicious - that fetch task handle should be preserved in the model in some way.

So yeah, I think looping should be the opportune way to handle that in a test environment - but maybe not with is_active() as I don't think that is set to false when the request completes - only when it is explicitly cancelled.

Maybe the best code would be something along the lines of:

```rust
  let mut stop = false;
  let callback = Callback::from(move |resp: Response<Result<String, failure::Error>>| {
     stop = true; // Not sure if the Callback's FnOnce bound will allow this to be satisfied
  });
  let task = FetchService::new().fetch_with_options(request, options, callback);
  while !stop {  }
```

Reply:
Nope, it doesnt allow it, unfortunately.

Reply:
Maybe use channels to circumvent the impossibility of setting a mut ref inside a callback:
```rust
use std::sync::mpsc::channel;

let (sender, receiver) = channel();

  let callback = Callback::from(move |resp: Response<Result<String, failure::Error>>| {
     sender.send()
  });
  let task = FetchService::new().fetch_with_options(request, options, callback);
  receiver.recv().unwrap() // wait for the callback to complete.
```

Reply:
Are they supported?
```
---- fetch_service::fetch_redirect_follow output ----
    error output:
        Encountered a panic!
        Panic location: src/libstd/sys/wasm/condvar.rs:23
    
    JS exception that was thrown:
        RuntimeError: unreachable
            at __rust_start_panic (wasm-function[6441]:0x15f5e3)
            at rust_panic (wasm-function[6232]:0x15d790)
            at std::panicking::rust_panic_with_hook::h5e7c2dc110ae79d4 (wasm-function[1351]:0xda358)
            at std::panicking::begin_panic::h837c5ed1e256db31 (wasm-function[6001]:0x15b243)
            at std::sys::wasm::condvar::Condvar::wait::hda825f3f4310d4db (wasm-function[6704]:0x160f48)
            at std::sys_common::condvar::Condvar::wait::h07e73ff239a34dad (wasm-function[6765]:0x16124d)
            at std::thread::park::h0838d30e3c3b47de (wasm-function[418]:0x845c2)
            at std::sync::mpsc::blocking::WaitToken::wait::haae7fa95d8f445ec (wasm-function[3705]:0x134537)
            at std::sync::mpsc::oneshot::Packet<T>::recv::h04e32325a509fe25 (wasm-function[192]:0x59a3d)
            at std::sync::mpsc::Receiver<T>::recv::h669ab51a2957fd08 (wasm-function[103]:0x361f8)
```

Reply:
I tried with an `Rc<RefCell<_>>` and it loops indefinitely.

```rust
use std::cell::RefCell;
use std::rc::Rc;
#[cfg(feature = "wasm_test")]
use wasm_bindgen_test::{wasm_bindgen_test as test, wasm_bindgen_test_configure};
use yew::callback::Callback;
use yew::format::Nothing;
use yew::services::fetch::{self, FetchOptions, Response};
use yew::services::FetchService;
use yew::services::Task;

#[cfg(feature = "wasm_test")]
wasm_bindgen_test_configure!(run_in_browser);

#[test]
fn fetch_redirect_follow() {
  let request = fetch::Request::get("https://httpbin.org/relative-redirect/1")
    .body(Nothing)
    .unwrap();
  let options = FetchOptions {
    redirect: Some(fetch::Redirect::Follow),
    ..FetchOptions::default()
  };
  let is_done = Rc::new(RefCell::new(false));

  let is_done_callback = is_done.clone();
  let callback = Callback::from(move |resp: Response<Result<String, failure::Error>>| loop {
    match is_done_callback.try_borrow_mut() {
      Ok(mut is_done_callback) => {
        *is_done_callback = true;
        return;
      }
      _ => {}
    }
  });
  let task = FetchService::new().fetch_with_options(request, options, callback);
  loop {
    match is_done.try_borrow() {
      Ok(is_done) if *is_done => return,
      _ => {}
    }
  }
}

```

Reply:
When I monitor traffic with Wireshark the request does happen though.

Reply:
I also tried with a full Yew app:

```rust
#[cfg(feature = "wasm_test")]
use wasm_bindgen_test::{wasm_bindgen_test as test, wasm_bindgen_test_configure};

use yew::format::Nothing;
use yew::services::fetch::{self, FetchOptions, FetchTask, Response};
use yew::services::FetchService;

#[cfg(feature = "wasm_test")]
wasm_bindgen_test_configure!(run_in_browser);

#[test]
fn fetch_redirect_follow() {
  use yew::{html, Component, ComponentLink, Html, ShouldRender};

  struct Model {
    link: ComponentLink<Self>,
    task: Option<FetchTask>,
  }
  enum Msg {
    TestSuccess,
    TestFail(Response<Result<String, failure::Error>>),
  }
  impl Component for Model {
    type Message = Msg;
    type Properties = ();
    fn create(_: Self::Properties, link: ComponentLink<Self>) -> Self {
      let mut model = Model { link, task: None };

      model.task = Some(
        FetchService::new().fetch_with_options(
          fetch::Request::get("https://httpbin.org/relative-redirect/1")
            .body(Nothing)
            .unwrap(),
          FetchOptions {
            redirect: Some(fetch::Redirect::Follow),
            ..FetchOptions::default()
          },
          model
            .link
            .callback(move |resp: Response<Result<String, failure::Error>>| {
              if resp.status().is_success() {
                Msg::TestSuccess
              } else {
                Msg::TestFail(resp)
              }
            }),
        ),
      );

      model
    }
    fn update(&mut self, msg: Self::Message) -> ShouldRender {
      match msg {
        Msg::TestSuccess => {
          self.task = None;
          false
        }
        Msg::TestFail(resp) => {
          self.task = None;
          panic!("{:?}", resp);
        }
      }
    }
    fn view(&self) -> Html {
      html! {}
    }
  }
  yew::start_app::<Model>();
}
```

And on test execution it returns:

```
Running headless tests in Chrome on `http://127.0.0.1:36599/`
Try find `webdriver.json` for configure browser's capabilities:
Not found
driver status: signal: 9                          
driver stdout:
    Starting ChromeDriver 78.0.3904.105 (60e2d8774a8151efa6a00b1f358371b1e0e07ee2-refs/branch-heads/3904@{#877}) on port 36599
    Only local connections are allowed.
    Please protect ports used by ChromeDriver and related test frameworks to prevent access by malicious code.

Error: invalid type: map, expected a string at line 1 column 68
error: test failed, to rerun pass '--test fetch_service'
```

Reply:
Evidently channels aren't supported.

I'm going to defer to @jstarry at this point as I don't have experience in writing tests in this context.

Reply:
It looks more supported to write asynchronous tests with asynchronous functions. https://rustwasm.github.io/docs/wasm-bindgen/wasm-bindgen-test/asynchronous-tests.html

Problem: Yew implements callback based services and not Futures.